dom-fetch-client/src/main/scala/org/http4s/dom/FetchClient.scala

```diff
@@ -20,30 +20,71 @@ package dom
 import cats.effect.Async
 import cats.effect.Resource
 import cats.syntax.all._
+import cats.effect.syntax.all._
+import java.util.concurrent.TimeoutException
 import org.http4s.client.Client
+import org.http4s.headers.{`Referer`}
 import org.scalajs.dom.crypto._
+import org.scalajs.dom.experimental.AbortController
 import org.scalajs.dom.experimental.Fetch
 import org.scalajs.dom.experimental.Headers
 import org.scalajs.dom.experimental.HttpMethod
 import org.scalajs.dom.experimental.RequestInit
+import org.scalajs.dom.experimental.{Response => FetchResponse}
+import scala.concurrent.duration._
 
 object FetchClient {
 
-  def apply[F[_]](implicit F: Async[F]): Client[F] = Client[F] { req: Request[F] =>
+  private[dom] def makeClient[F[_]](
+      requestTimeout: Duration,
+      options: FetchOptions
+  )(implicit F: Async[F]): Client[F] = Client[F] { (req: Request[F]) =>
     Resource.eval(req.body.chunkAll.filter(_.nonEmpty).compile.last).flatMap { body =>
       Resource
         .makeCase {
-          F.fromPromise {
-            F.delay {
-              val init = new RequestInit {}
+          F.delay(new AbortController()).flatMap { abortController =>
+            val requestOptions = req.attributes.lookup(FetchOptions.Key)
+            val mergedOptions = requestOptions.fold(options)(options.overrideWith)
```

Comment:
Regarding the `AbortController`: I was imagining something more like this.
```scala
Resource.makeCase(F.delay(new AbortController)) {
  case (controller, Resource.ExitCase.Succeeded) => F.unit
  case (controller, Resource.ExitCase.Error(ex)) => F.delay(r.abort())
  case (controller, Resource.ExitCase.Cancelled) => F.delay(r.abort())
}.flatMap { abortController =>
  Resource.makeCase(/* fetch code in here */) { case (r, exitCase) =>
    closeReadableStream(r.body, exitCase)
  }
}.timeoutTo(d, F.raiseError(/* etc */))
```

Reply:
Hmm, on second thought I see that this is flawed, sorry. 🤔

Reply:
Ok, I think there's still some trickyness here. So in `Resource.makeCase(acquire)(release)` acquire is `atomic` i.e. uncancelable. So even if the request is cancelled externally, it will still block until the request completes/times out. I think we want to use `makeCaseFull` and wrap the entirety of `acquire` inside of the `poll`, because it _is_ cancelable thanks to your addition of the `AbortController`.

https://typelevel.org/cats-effect/docs/typeclasses/monadcancel

Reply:
That's a very good catch. I admit I forgot about atomic acquires. I'll look into it.